prov/efa/src/efa_cq.c

```diff
@@ -0,0 +1,535 @@
+			switch (opcode) {
+			case IBV_WC_SEND: /* fall through */
+			case IBV_WC_RDMA_WRITE: /* fall through */
+			case IBV_WC_RDMA_READ:
+				efa_cq_handle_error(base_ep, cq->ibv_cq.ibv_cq_ex,
+						    base_ep->util_ep.tx_cq,
+						    to_fi_errno(prov_errno),
+						    prov_errno);
+				break;
+			case IBV_WC_RECV: /* fall through */
+			case IBV_WC_RECV_RDMA_WITH_IMM:
+				if (efa_cq_wc_is_unsolicited(cq->ibv_cq.ibv_cq_ex)) {
+					EFA_WARN(FI_LOG_CQ,
+						 "Receive error %s (%d) for unsolicited write recv",
+						 efa_strerror(prov_errno), prov_errno);
+					efa_base_ep_write_eq_error(
+						base_ep,
+						to_fi_errno(prov_errno),
+						prov_errno);
+					break;
+				}
+				efa_cq_handle_error(base_ep, cq->ibv_cq.ibv_cq_ex,
+						    base_ep->util_ep.rx_cq,
+						    to_fi_errno(prov_errno),
+						    prov_errno);
+				break;
+			default:
+				EFA_WARN(FI_LOG_EP_CTRL, "Unhandled op code %d\n", opcode);
+				assert(0 && "Unhandled op code");
+			}
+			break;
+		}
+
+		cq->read_entry(cq->ibv_cq.ibv_cq_ex, &cq_entry);
+
+		switch (opcode) {
+		case IBV_WC_SEND: /* fall through */
+		case IBV_WC_RDMA_WRITE: /* fall through */
+		case IBV_WC_RDMA_READ:
+			efa_cq_handle_send_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex, &cq_entry);
+			break;
+		case IBV_WC_RECV:
+			efa_cq_handle_recv_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex, &cq_entry);
+			break;
+		case IBV_WC_RECV_RDMA_WITH_IMM:
+			efa_cq_proc_ibv_recv_rdma_with_imm_completion(
+				base_ep, cq->ibv_cq.ibv_cq_ex);
+			break;
+		default:
+			EFA_WARN(FI_LOG_EP_CTRL,
+				"Unhandled cq type\n");
+			assert(0 && "Unhandled cq type");
+		}
+
+		num_cqe++;
+		if (num_cqe == cqe_to_process) {
+			break;
+		}
+
+		err = ibv_next_poll(cq->ibv_cq.ibv_cq_ex);
+	}
+
+	if (err && err != ENOENT) {
+		err = err > 0 ? err : -err;
+		prov_errno = ibv_wc_read_vendor_err(cq->ibv_cq.ibv_cq_ex);
+		EFA_WARN(FI_LOG_CQ,
+			 "Unexpected error when polling ibv cq, err: %s (%zd) "
+			 "prov_errno: %s (%d)\n",
+			 fi_strerror(err), err, efa_strerror(prov_errno),
+			 prov_errno);
+		efa_show_help(prov_errno);
+		err_entry = (struct fi_cq_err_entry) {
+			.err = err,
+			.prov_errno = prov_errno,
+			.op_context = NULL,
+		};
+		ofi_cq_write_error(&cq->util_cq, &err_entry);
+	}
+
+	if (should_end_poll)
+		ibv_end_poll(cq->ibv_cq.ibv_cq_ex);
+}
+
+static const char *efa_cq_strerror(struct fid_cq *cq_fid,
+				   int prov_errno,
+				   const void *err_data,
+				   char *buf, size_t len)
+{
+	return err_data
+		? (const char *) err_data
+		: efa_strerror(prov_errno);
+}
+
+static struct fi_ops_cq efa_cq_ops = {
+	.size = sizeof(struct fi_ops_cq),
+	.read = ofi_cq_read,
+	.readfrom = ofi_cq_readfrom,
+	.readerr = ofi_cq_readerr,
+	.sread = fi_no_cq_sread,
+	.sreadfrom = fi_no_cq_sreadfrom,
+	.signal = fi_no_cq_signal,
+	.strerror = efa_cq_strerror
+};
+
+void efa_cq_progress(struct util_cq *cq)
+{
+	efa_cq_poll_ibv_cq(efa_env.efa_cq_read_size, cq);
```

Comment:
As I mentioned earlier, we may need a domain level lock here to protect the `qp_table` which is also accessed by the `ep_enable` and `ep_close` calls. But I am ok with making it as part of a follow up change when we implement the `ep` interface for efa raw.